src/orange/orangeinp/detail/DeMorganSimplifier.cc

```diff
@@ -0,0 +1,443 @@
+                            // parent for that node
+                            CELER_EXPECT(node_ids_translation_[n].modified);
+                            operands.push_back(
+                                *node_ids_translation_[n].modified);
+                        }
+                    }
+
+                    auto [new_id, inserted] = result.insert(
+                        Joined{(op == logic::land) ? logic::lor : logic::land,
+                               std::move(operands)});
+                    node_ids_translation_[node_id].modified = std::move(new_id);
+                }
+                // if this joined node doesn't have any
+                // parents other than the simplified negation, we don't need to
+                // insert
+                return !(orphaned_nodes_[node_id.get()]);
+            },
+            // other nodes need to be inserted
+            [](auto&&) { return true; },
+        },
+        tree_[node_id]);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * First pass through the tree to find negated set operations.
+ */
+void DeMorganSimplifier::find_join_negations()
+{
+    for (auto node_id : range(NodeId{tree_.size()}))
+    {
+        std::visit(Overload{
+                       [&](Negated const& negated) {
+                           // record_parents unmark a Joined or Negated node
+                           // previously marked as orphan, so it doesn't need
+                           // to be called here because a child of a Negated
+                           // node always simplifies (double negation or
+                           // negated join)
+                           if (std::get_if<Joined>(&tree_[negated.node]))
+                           {
+                               // This is a Negated{Joined{...}}
+                               record_join_negation(node_id);
+                           }
+                       },
+                       [&](Joined const& joined) {
+                           // we found a new parent for each operand
+                           for (auto const& join_operand : joined.nodes)
+                           {
+                               record_parent_for(join_operand);
+                           }
+                       },
+                       [&](Aliased const& aliased) {
+                           record_parent_for(aliased.node);
+                       },
+                       // nothing to do for leaf node types
+                       [](auto&&) {},
+                   },
+                   tree_[node_id]);
+    }
+
+    // Volume nodes act as tags on a NodeId indicating that it is the root of a
+    // volume, so these subtrees need to be preserved.
+    // Consider a "virtual" parent for these nodes
+    std::unordered_set<NodeId> volume_roots{tree_.volumes().cbegin(),
+                                            tree_.volumes().cend()};
+    for (auto node_id : volume_roots)
```

Comment:
No two volumes should point to the same node: that implies two volumes occupy the same exact region of space, which is a geometry error. We could in theory check this in the CSG tree itself. For now, I think the code will still work even if the same NodeId is shared by two volumes...
```suggestion
    for (auto node_id : tree_.volumes())
```